test/e2e/k8s_util.go

```diff
@@ -0,0 +1,453 @@
+	})
+	if err != nil {
+		return nil, errors.WithMessage(err, "unable to list pods")
+	}
+	return v1PodList.Items, nil
+}
+
+// GetPods returns an array of all pods in the given namespace having a k/v label pair.
+func (k *KubernetesUtils) GetPods(ns string, key string, val string) ([]v1.Pod, error) {
+	if p, ok := k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)]; ok {
+		return p, nil
+	}
+
+	v1PodList, err := k.getPodsUncached(ns, key, val)
+	if err != nil {
+		return nil, errors.WithMessage(err, "unable to list pods")
+	}
+	k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)] = v1PodList
+	return v1PodList, nil
+}
+
+// Probe execs into a pod and checks its connectivity to another pod.  Of course it assumes
+// that the target pod is serving on the input port, and also that wget is installed.  For perf it uses
+// spider rather then actually getting the full contents.
+func (k *KubernetesUtils) Probe(ns1, pod1, ns2, pod2 string, port int) (bool, error) {
+	fromPods, err := k.GetPods(ns1, "pod", pod1)
+	if err != nil {
+		return false, errors.WithMessagef(err, "unable to get pods from ns %s", ns1)
+	}
+	if len(fromPods) == 0 {
+		return false, errors.New(fmt.Sprintf("no pod of name %s in namespace %s found", pod1, ns1))
+	}
+	fromPod := fromPods[0]
+
+	toPods, err := k.GetPods(ns2, "pod", pod2)
+	if err != nil {
+		return false, errors.WithMessagef(err, "unable to get pods from ns %s", ns2)
+	}
+	if len(toPods) == 0 {
+		return false, errors.New(fmt.Sprintf("no pod of name %s in namespace %s found", pod2, ns2))
+	}
+	toPod := toPods[0]
+
+	toIP := toPod.Status.PodIP
+
+	// There seems to be an issue when running Antrea in Kind where tunnel traffic is dropped at
+	// first. This leads to the first test being run consistently failing. To avoid this issue
+	// until it is resolved, we try to connect 3 times.
+	// See https://github.com/vmware-tanzu/antrea/issues/467.
+	cmd := []string{
+		"/bin/sh",
+		"-c",
+		// 3 tries, timeout is 1 second
+		fmt.Sprintf("for i in $(seq 1 3); do ncat -vz -w 1 %s %d && exit 0 || true; done; exit 1", toIP, port),
+	}
+	// HACK: inferring container name as c80, c81, etc, for simplicity.
+	containerName := fmt.Sprintf("c%v", port)
+	log.Tracef("Running: kubectl exec %s -c %s -n %s -- %s", fromPod.Name, containerName, fromPod.Namespace, strings.Join(cmd, " "))
+	stdout, stderr, err := k.runCommandFromPod(fromPod.Namespace, fromPod.Name, containerName, cmd)
+	if err != nil {
+		// log this error as trace since may be an expected failure
+		log.Tracef("%s/%s -> %s/%s: error when running command: err - %v /// stdout - %s /// stderr - %s", ns1, pod1, ns2, pod2, err, stdout, stderr)
+		// do not return an error
+		return false, nil
+	}
+	return true, nil
+}
+
+// ExecuteRemoteCommand executes a remote shell command on the given pod
+// returns the output from stdout and stderr
+func (k *KubernetesUtils) ExecuteRemoteCommand(pod v1.Pod, cname string, command []string) (string, string, error) {
```

Comment:
why don't you use `k. runCommandFromPod` since `KubernetesUtils` embeds `TestData`?

Reply:
This function was ported as part of the original framework and is now stale. It is actually not being called anywhere since the traffic tests are evoked through `k.runCommandFromPod` already. Removed this.